spring-ai-core/src/main/java/org/springframework/ai/client/metadata/Usage.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright 2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.ai.client.metadata;
+
+/**
+ * Abstract Data Type (ADT) encapsulating metadata on the usage of an AI provider's API
+ * per AI request.
+ *
+ * @author John Blum
+ * @since 0.7.0
+ */
+public interface Usage {
```

Comment:
I think it is possible to go with the LCD approach that is type safe, though we need more investigation. 

Here is some analysis.  The huggingface one is the most different since it supports so many models.

One thing i thought was to provide some sort of type-safe client-specific helper class that given a hashmap would return typesafe results for devs who want to access the raw return information almost as if they were using the vendor specific API.



 For example, here is the 'Details' class from the hugging face client.
```
  @JsonProperty("best_of_sequences")
  private List<BestOfSequence> bestOfSequences = null;

  @JsonProperty("finish_reason")
  private FinishReason finishReason = null;

  @JsonProperty("generated_tokens")
  private Integer generatedTokens = null;

  @JsonProperty("prefill")
  private List<PrefillToken> prefill = new ArrayList<>();

  @JsonProperty("seed")
  private Long seed = null;

  @JsonProperty("tokens")
  private List<Token> tokens = new ArrayList<>();
```

It is interesting they dont' have the separation into prompt tokens and tokens that were generated, though one might be able to calculate the prompt tokens in our own client and then do the subtraction from the total to get at the generated token value (estimate).

and for the azure open ai client 

```
@Immutable
public final class CompletionsUsage {

    /*
     * The number of tokens generated across all completions emissions.
     */
    @Generated
    @JsonProperty(value = "completion_tokens")
    private int completionTokens;

    /*
     * The number of tokens in the provided prompts for the completions request.
     */
    @Generated
    @JsonProperty(value = "prompt_tokens")
    private int promptTokens;

    /*
     * The total number of tokens processed for the completions request and response.
     */
    @Generated
    @JsonProperty(value = "total_tokens")
    private int totalTokens;
```

And from the theo kanning open ai client



```
public class Usage {
    @JsonProperty("prompt_tokens")
    long promptTokens;
    @JsonProperty("completion_tokens")
    long completionTokens;
    @JsonProperty("total_tokens")
    long totalTokens;
```

some

Reply:
Agreed, and an excellent point. I will investigate on this topic more. Thank you for the feedback and references.